### ISSUE TITLE
Mapping interface hierarchy

### ISSUE THREAD

**Issue by Divisibility** [modelmapper/modelmapper#421]
Hi there,

im having difficulties to use ModelMapper to map our JPA-Entity tree to our TO object tree. We are working with TO-Java-Interfaces and are hiding implementations of TOs to our service-class. This class maps the entities by using self-written converters classes. These converters are instantiating correct implementation of TO-Interface and are filling the data. The converters are using other converters to map the whole object hierarchy.

W'd like to exchange our converters with ModelMapper and I tried to use ModelMapper within one of our services:

Here is my interface TO setup:
ISampleTO has ISub1TO and ISub2TO. ISub1TO has ISub3TO.

```
interface ISampleTO {
	Long getId();
	void setId(Long id);

	ISub1TO getSub1TO();
	void setSub1TO(ISub1TO id);

	ISub2TO getSub2TO();
	void setSub2TO(ISub2TO id);
	...	
}

interface ISub1TO {
	Long getId();
	void setId(Long id);

	ISub3TO getSub3TO();
	void setSub3TO(ISub3TO id);
	...	
}

interface ISub2TO {
	Long getId();
	void setId(Long id);
	...
}

interface ISub3TO {
	Long getId();
	void setId(Long id);
	...	
}
```

same hierarchy with SampleEntity, but with Entity instead of TO in name.

Service class has private Methods to encapsulate ModelMapper:

```
private SampleEntity createFromDto(ISampleTO to) {
    return modelMapper.map(to, SampleEntity.class);
}

private ISampleTO createFromEntity(SampleEntity entity) {

    ISampleTO newTo = new SampleTO();
    modelMapper.map(entity, newTo);
    return newTo;
}
```

When I try to run this I'm getting the following error from ModelMapper:
```
Failed to instantiate instance of destination ISub1TO. Ensure that ISub1TO has a non-private no-argument constructor.
```

I found issue https://github.com/modelmapper/modelmapper/issues/13 but as far as I understand ModelMapper does not support mapping of nested interface-TOs. Is this correct? The last post refers to "TypeMap inheritance", but when I read the documentation I can not see any helpful information to a solution of my problem.

My questions are:
- Can ModelMapper map a hierarchy of interfaces?
- If yes, how can I tell ModelMapper which realization it should be used to map to (interfaces can have more than one implementation)
- If yes, how would this be achieved, I don't have a clue.

Thank you very much.

**Comment by chhsiao90** [comment]
Hi @Divisibility ,

We have [Type Map Inheritance](http://modelmapper.org/user-manual/type-map-inheritance/) that was used for solving this kind of problems.

I guess your entity will looks like below.
```java
class SampleEntity {
  Long id;
  Sub1 sub1;
  Sub2 sub2;
  Sub3 sub3;
}
```

And you need to configure your ModelMapper as below.
```java
modelMapper.typeMap(Sub1.class, Sub1TO.class).include(ISub1TO.class);
modelMapper.typeMap(Sub2.class, Sub2TO.class).include(ISub2TO.class);
modelMapper.typeMap(Sub3.class, Sub3TO.class).include(ISub3TO.class);
```

**Comment by Divisibility** [comment]
Hi @chhsiao90,

I have tested your suggestion and unfortantely it does not work. Im getting the same errors:
  
  1) Failed to instantiate instance of destination me.dames.to.ISub1TO. Ensure that me.dames.to.ISub1TO has a non-private no-argument constructor.
  Caused by: java.lang.NoSuchMethodException: me.dames.to.ISub1TO.<init>()
  ..
  2) Failed to instantiate instance of destination me.dames.to.ISub1TO. Ensure that me.dames.to.ISub1TO has a non-private no-argument constructor.
  Caused by: java.lang.NoSuchMethodException: me.dames.to.ISub1TO.<init>()
  ..
  3) Failed to instantiate instance of destination me.dames.to.ISub2TO. Ensure that me.dames.to.ISub2TO has a non-private no-argument constructor.
  Caused by: java.lang.NoSuchMethodException: me.dames.to.ISub2TO.<init>()
  ..

I have attached my sample project as [zip](https://github.com/modelmapper/modelmapper/files/2698810/mapInterfacesTOs.zip) and I'd appreciate some more assistence a lot! 

Thank you.
